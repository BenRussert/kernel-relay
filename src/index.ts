import { createServer, kernels } from "./server";

async function main() {
  // In the most basic sense, the ApolloServer can be started
  // by passing type definitions (typeDefs) and the resolvers
  // responsible for fetching the data for those types.
  const server = createServer();
  // This `listen` method launches a web-server.  Existing apps
  // can utilize middleware options, which we'll discuss later.
  server.listen().then(({ url }) => {
    console.log(`ðŸš€  Server ready at ${url}`);
  });
}

process.on("exit", () => {
  Object.keys(kernels).map(async id => {
    console.log("shutting down ", id);
    await kernels[id].shutdown();
  });
});

main();
