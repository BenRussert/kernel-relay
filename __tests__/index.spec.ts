import { gql } from "apollo-server";
import { createTestClient } from "apollo-server-testing";
import { createServer } from "../src/server";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "kernel1")
}));

jest.mock("@nteract/fs-kernels", () => ({
  async findAll() {
    return {
      python3: {
        name: "python3",
        files: [
          "/Users/jovyan/Library/Jupyter/kernels/python3/kernel.json",
          "/Users/jovyan/Library/Jupyter/kernels/python3/logo-32x32.png",
          "/Users/jovyan/Library/Jupyter/kernels/python3/logo-64x64.png"
        ],
        resource_dir: "/Users/jovyan/Library/Jupyter/kernels/python3",
        spec: {
          argv: [
            "/usr/local/bin/python3",
            "-m",
            "ipykernel_launcher",
            "-f",
            "{connection_file}"
          ],
          env: {},
          display_name: "Python 3",
          language: "python",
          interrupt_mode: "signal"
        }
      }
    };
  },
  async launchKernel(name: string) {
    return {
      connectionInfo: {
        key: "kernel1"
      },
      shutdown: jest.fn(),
      channels: {
        next: jest.fn(),
        subscribe: jest.fn(),
        unsubscribe: jest.fn(),
        complete: jest.fn()
      }
    };
  }
}));

describe("Queries", () => {
  it("returns a list of kernelspecs", async () => {
    const { query } = createTestClient(createServer());
    const LIST_KERNELSPECS = gql`
      query GetKernels {
        listKernelSpecs {
          name
        }
      }
    `;
    const response = await query({ query: LIST_KERNELSPECS });
    expect(response).toMatchSnapshot();
  });
});

describe("Mutations", () => {
  let kernelId;
  it("launches a kernel", async () => {
    const { mutate } = createTestClient(createServer());
    const START_KERNEL = gql`
      mutation StartJupyterKernel {
        startKernel(name: "python3") {
          id
          status
        }
      }
    `;

    const response = await mutate({ mutation: START_KERNEL });

    // When the response has errors, they'll be an array
    // and it's easier to diagnose if we use an expect matcher here
    expect(response.errors).toBeUndefined();

    kernelId = response.data.startKernel.id;

    expect(response).toMatchSnapshot();
  });
  it("shuts down a kernel", async () => {
    const { mutate } = createTestClient(createServer());
    const SHUTDOWN_KERNEL = gql`
      mutation KillJupyterKernel($id: ID) {
        shutdownKernel(id: $id) {
          id
          status
        }
      }
    `;
    const response = await mutate({
      mutation: SHUTDOWN_KERNEL,
      variables: { id: kernelId }
    });

    // When the response has errors, they'll be an array
    // and it's easier to diagnose if we use an expect matcher here
    expect(response.errors).toBeUndefined();

    expect(response).toMatchSnapshot();
  });
});
