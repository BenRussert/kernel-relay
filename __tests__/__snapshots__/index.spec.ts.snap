// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutations launches a kernel 1`] = `
Object {
  "data": Object {
    "startKernel": Object {
      "id": "kernel1",
      "status": "launched",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutations shuts down a kernel 1`] = `
Object {
  "data": Object {
    "shutdownKernel": Object {
      "id": "kernel1",
      "status": "shutdown",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries returns a list of kernelspecs 1`] = `
Object {
  "data": Object {
    "listKernelSpecs": Array [
      Object {
        "name": "python3",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
